File: games/winter2026/engine/command_manager.go
// Package winter2026
// Source: WinterChallenge2026-Exotec/src/main/java/com/codingame/game/CommandManager.java
package engine

import (
	"fmt"
	"strings"

	"github.com/mrsombre/codingame-arena/games/winter2026/engine/action"
)

type CommandManager struct {
	summary *[]string
}

func NewCommandManager(summary *[]string) *CommandManager {
	return &CommandManager{summary: summary}
}

func (m *CommandManager) ParseCommands(player *Player, lines []string) {
	if len(lines) == 0 {
		m.deactivatePlayer(player, "Timeout!")
		player.SetTimedOut(true)
		return
	}

	errors := make([]string, 0)
	commands := splitCommands(lines[0])
	reasonableLimitForActions := 30
	for _, command := range commands {
		if reasonableLimitForActions <= 0 {
			return
		}
		reasonableLimitForActions--

		parsed, err := action.Parse(command)
		if err != nil {
			inputErr := &InvalidInputError{Expected: GetExpected(command), Got: command}
			m.deactivatePlayer(player, inputErr.Error())
			m.addSummary(inputErr.Error())
			m.addSummary(fmt.Sprintf("$%d: disqualified!", player.GetIndex()))
			break
		}

		if parsed.IsMove() {
			birdID := parsed.BirdID
			bird := player.BirdByID(birdID)
			switch {
			case bird == nil:
				errors = append(errors, formatError(fmt.Sprintf("$%d: Bird not found for id %d", player.GetIndex(), birdID)))
				continue
			case !bird.Alive:
				errors = append(errors, formatError(fmt.Sprintf("$%d: Bird with id %d is dead", player.GetIndex(), birdID)))
				continue
			case bird.HasMove:
				errors = append(errors, formatError(fmt.Sprintf("$%d: Bird id %d has already been given a move", player.GetIndex(), birdID)))
				continue
			case bird.Facing().Opposite() == parsed.Direction:
				errors = append(errors, formatError(fmt.Sprintf("$%d: Bird id %d cannot move backwards", player.GetIndex(), birdID)))
				continue
			}
			bird.Direction = parsed.Direction
			bird.HasMove = true
			if parsed.HasMessage {
				bird.SetMessage(parsed.Message)
			}
		} else if parsed.IsMark() {
			if !player.AddMark(parsed.Coord) {
				errors = append(errors, formatError(fmt.Sprintf("$%d: Too many MARK actions this turn", player.GetIndex())))
			}
		}
	}

	maxErrs := 4
	if len(errors) <= maxErrs+1 {
		for _, err := range errors {
			m.addSummary(err)
		}
	} else {
		for _, err := range errors[:maxErrs] {
			m.addSummary(err)
		}
		m.addSummary(formatError(fmt.Sprintf("...and %d more errors.", len(errors)-maxErrs)))
	}
}

func (m *CommandManager) deactivatePlayer(player *Player, message string) {
	player.Deactivate(escapeHTMLEntities(message))
	player.SetScore(-1)
}

func (m *CommandManager) addSummary(line string) {
	if m.summary != nil {
		*m.summary = append(*m.summary, line)
	}
}

func escapeHTMLEntities(message string) string {
	return strings.NewReplacer("&lt;", "<", "&gt;", ">").Replace(message)
}

func formatError(message string) string {
	return message
}

func splitCommands(line string) []string {
	commands := strings.Split(line, ";")
	for len(commands) > 1 && commands[len(commands)-1] == "" {
		commands = commands[:len(commands)-1]
	}
	return commands
}

File: source/SummerChallenge2026-BackTrackKing/src/main/java/com/codingame/game/CommandManager.java
package com.codingame.game;

import java.util.List;
import java.util.Optional;
import java.util.regex.Matcher;

import com.codingame.game.action.Action;
import com.codingame.game.action.ActionType;
import com.codingame.gameengine.core.GameManager;
import com.codingame.gameengine.core.MultiplayerGameManager;
import com.google.inject.Inject;
import com.google.inject.Singleton;

@Singleton
public class CommandManager {

    @Inject private MultiplayerGameManager<Player> gameManager;

    public void parseCommands(Player player, List<String> lines) {

        String line = lines.get(0);
        String[] commands = line.trim().split(";");
        try {
            Matcher match;
            try {
                for (String _command : commands) {
                    String command = _command.trim();
                    boolean found = false;
                    for (ActionType actionType : ActionType.values()) {

                        match = actionType.getPattern().matcher(command);
                        if (match.matches()) {
                            Action action = new Action(actionType);
                            actionType.getConsumer().accept(match, action);

                            if (action.getType() == ActionType.MESSAGE) {
                                player.setMessage(action.getMessage());
                            } else {
                                player.intents.add(action);
                            }

                            found = true;
                        }
                        if (found) {
                            break;
                        }
                    }
                    if (!found) {
                        throw new InvalidInputException(Game.getExpected(command), command);
                    }
                }
            } catch (Exception e) {
                throw e;
            }

        } catch (Exception e) {
            deactivatePlayer(player, e.getMessage());
            gameManager.addToGameSummary(e.getMessage());
            gameManager.addToGameSummary(GameManager.formatErrorMessage(player.getNicknameToken() + ": disqualified!"));
        }

    }

    private Optional<Integer> getAgentId(String[] tokens) {
        if (tokens.length > 1) {
            try {
                return Optional.of(Integer.parseInt(tokens[0]));
            } catch (NumberFormatException e) {
                return Optional.empty();
            }
        }
        return Optional.empty();
    }

    public void deactivatePlayer(Player player, String message) {
        player.deactivate(escapeHTMLEntities(message));
        player.setScore(-1);
    }

    private String escapeHTMLEntities(String message) {
        return message
            .replace("&lt;", "<")
            .replace("&gt;", ">");
    }
}
